File: src/main/kotlin/tracking/website/Strategy.kt
package tracking.website

import web.url.Url

interface Strategy {
    fun isSuitableFor(url: String): Boolean
    fun filterContents(input: String): String = input
    fun getUrlForTracking(url: Url): Url = url
    fun getCookieInitUrl(): Url? = null
    fun getLatentCode(url: Url, contents: String, originalCode: Int): Int = originalCode

    companion object {
        fun forUrl(url: String): Strategy {
            return when {
                TwitterProfileStrategy
                    .isSuitableFor(url) -> TwitterProfileStrategy

                InstagramProfileStrategy
                    .isSuitableFor(url) -> InstagramProfileStrategy

                TrelloStrategy
                    .isSuitableFor(url) -> TrelloStrategy

                FurAffinityProfileStrategy
                    .isSuitableFor(url) -> FurAffinityProfileStrategy

                else -> StandardStrategy
            }
        }
    }
}

File: symfony/src/Utils/Web/UrlStrategy/Strategy.php
<?php

declare(strict_types=1);

namespace App\Utils\Web\UrlStrategy;

use App\Utils\Web\Url\Url;
use Composer\Pcre\Preg;

abstract class Strategy
{
    // Seems to duplicate primary content on some serial websites.
    private const string META_DESCRIPTION_PATTERN = '~<meta (?:itemprop|property)="(?:og:|twitter:)?description"[^>]+>~';

    abstract public static function isSuitableFor(string $url): bool;

    public function filterContents(string $input): string
    {
        return Preg::replace(self::META_DESCRIPTION_PATTERN, '', $input);
    }

    public function getUrlForTracking(Url $url): Url
    {
        return $url;
    }

    public function getCookieInitUrl(): ?Url
    {
        return null;
    }

    public function getLatentCode(Url $url, string $contents, int $originalCode): int
    {
        return $originalCode;
    }
}
